### ISSUE TITLE
SCTP channel issue in RHEL 8.2

### ISSUE THREAD

**Issue by LosAngelesHK** [netty/netty#10455]
![SCTP-Packet-Trace](https://user-images.githubusercontent.com/13230148/89488931-dcc61480-d7db-11ea-82d1-8c038f032c7a.jpg)
### Expected behavior
When the SCTP connection is established successfully, channelRegistered() method and channelActive() method of channel inbound handler should be invoked in the SCTP server application.

### Actual behavior
When the SCTP connection is established successfully, only channelRegistered() method of channel inbound handler is invoked and channelActive() method has not triggered to invoke

### Steps to reproduce
1. Deploy the SCTP server application in RHEL 8.2
2. Start the SCTP server application
3. SCTP server application is listened on 10.165.238.39:3868 successfully
4. Trigger a SCTP connection to the SCTP server application
5. In the packet trace, the SCTP connection is handshaked and established successfully
6. In the SCTP server application, channelRegistered() method is invoked
7. However, channelActive() method has not triggered to invoke
p.s. The following options have been used when starting the SCTP server application:
      -Dio.netty.leakDetectionLevel=advanced
      -Dio.netty.tryReflectionSetAccessible=true
      --add-opens java.base/jdk.internal.misc=ALL-UNNAMED
      --illegal-access=warn

### Minimal yet complete reproducer code (or URL to code)
N/A

### Netty version
4.1.51.Final

### JVM version (e.g. `java -version`)
openjdk version "14.0.2" 2020-07-14
OpenJDK Runtime Environment (build 14.0.2+12-46)

### OS version (e.g. `uname -a`)
Linux test-server 4.18.0-193.13.2.el8_2.x86_64 #1 SMP Mon Jul 13 23:17:28 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

**Comment by normanmaurer** [comment]
@LosAngelesHK can you also reproduce this with prior JDK versions ?

**Comment by LosAngelesHK** [comment]
> @LosAngelesHK can you also reproduce this with prior JDK versions ?

Yes, JDK 8 and JDK 11 also have the same issue and behaviour

**Comment by normanmaurer** [comment]
@LosAngelesHK did you verify that it works with "stock" JDK implementation ? We basically just delegate to the JDK implementation

**Comment by LosAngelesHK** [comment]
> @LosAngelesHK did you verify that it works with "stock" JDK implementation ? We basically just delegate to the JDK implementation

@normanmaurer Sorry, I do not fully understand your question. What is your meaning of "stock" JDK implementation? BTW, the SCTP server application works normally using JDK 8/14 under the RHEL 6 & RHEL 7 environment

**Comment by LosAngelesHK** [comment]
@normanmaurer I have tried to compile and execute the SCTP server application using Oracle JDK 8 and Oracle JDK 14 in RHEL 8 environment, however both have the same issue and behaviour as OpenJDK

**Comment by normanmaurer** [comment]
@LosAngelesHK honestly I have no idea ... We just use the SCTP implementation provided by the JDK. So I guess the best would be to open an issue in the openjdk issue tracker

**Comment by normanmaurer** [comment]
Or maybe it is related to the running kernel ?

**Comment by LosAngelesHK** [comment]
@normanmaurer I found that there is a difference in the behaviour of "com.sun.nio.sctp.SctpChannel" between RHEL 7 and RHEL 8. Its method association() always returns null in RHEL 8 while RHEL 7 is not. ttAs I know, the method association() of "NioSctpChannel" in Netty is also invoking the same method of "com.sun.nio.sctp.SctpChannel". Is such behavioural difference related to the issue without channelActive()?

Attached is the simple SCTP client/server program to test the behaviour of "com.sun.nio.sctp.SctpChannel".
[TestSctpCliSvr.zip](https://github.com/netty/netty/files/5133647/TestSctpCliSvr.zip)

**Comment by normanmaurer** [comment]
hmm... that's interesting. this explains it as a channel is only treaded as active when `association() != null`.

**Comment by LosAngelesHK** [comment]
@normanmaurer I have just tried an experiment. Create a dummy Association object and return it in the association() method of class "NioSctpChannel". channelActive() method is invoked successfully in the SCTP server application.

**Comment by LosAngelesHK** [comment]
@normanmaurer After further testing the behaviour of "com.sun.nio.sctp.SctpChannel", I found that its association() method would NOT return null if I invoke SctpChannel.receive() method first. However, I think it may violate the mechanism of channel establishment in Netty framework ( channelRegistered() -> channelActive() -> channelRead() ). Do you have any idea about fixing the issue?

Attached is the modified version of above simple SCTP client/server program which is invoked the SctpChannel.receive() method first.
[TestSctpCliSvrNew.zip](https://github.com/netty/netty/files/5148197/TestSctpCliSvrNew.zip)

**Comment by normanmaurer** [comment]
@LosAngelesHK not really ... Imho this is either a CentOS bug or a JDK bug.

**Comment by poekmorgan** [comment]
@LosAngelesHK 

Sorry to step in. I read through this case. it sounds like the problem happens when running SCTP transport on Netty in kernel 4.x?

And, the JDK "com.sun.nio.sctp.SctpChannel" is working in kernel 2.6 or 3.x, but not in 4.x. is there any JDK function call can settle the problem? 

@normanmaurer , netty "NioSctpChannel" likely calls the "com.sun.nio.sctp.SctpChannel" to establish the SCTP stream, so it hits the problem. If there is a resolution using other function/API in JDK to settle the issue, seems like Netty is also needed to support. Otherwise, it will not work in kernel 4.x,

my 2 c only.

-pm

**Comment by LosAngelesHK** [comment]
@normanmaurer @poekmorgan 
> Sorry to step in. I read through this case. it sounds like the problem happens when running SCTP transport on Netty in kernel 4.x?

Yes.

> And, the JDK "com.sun.nio.sctp.SctpChannel" is working in kernel 2.6 or 3.x, but not in 4.x. the problem is whether any JDK function call can settle the problem? if yes, can you show the details?

Actually, I have already contacted Red Hat about this issue and they replied me that invoking SctpChannel.receive() method could be a workaround (Demonstrated in above reply with the sample client/server program "TestSctpCliSvrNew.zip"). As a result, I have tried to modify the code in "NioSctpChannel" as below:

    @Override
    public Association association() {
      try {
        Association assoc = javaChannel().association();
        if ( ( assoc == null ) && ( parent() != null ) ) {
          ByteBuffer buf = ByteBuffer.allocate( 50 );
          javaChannel().receive( buf, null, null );
          assoc = javaChannel().association();
        }
        return assoc;
      } catch ( IOException ignored ) {
        return null;
      }
    }
Finally, NioSctpChannel could enter the channelActive() and channelRead() states successfully if the returned association object is not null. However, I found that there is a side effect in such modification. Some data in received SCTP payload has been lost when reading the first SCTP message under the state of channelRead(). The lost data is due to the invoking of receive() method when retrieving the association object.

In conclusion, there is a difficulty in both sides. If I do not invoke the receive() method, then the retrieved association object is always null and thus the state of SctpChannel is blocked at channelRegistered() forever. On the other hand, if I invoke the receive() method, the state of SctpChannel could enter channelActive() and channelRead() successfully, but some data in the payload of first SCTP message has been lost.

**Comment by pgrabowski** [comment]
We're having similar issue but with Netty SCTP client not getting notification that SCTP association was created.
This is also RHEL 8.2 and JDK 8 from Oracle.
For us downgrading kernel to version from RHEL 8.0 solved this issue.
We are currently trying to pin point release where issue was introduced.

**Comment by poekmorgan** [comment]
@normanmaurer do you think there is something that Netty can help?

**Comment by pgrabowski** [comment]
> @pgrabowski do you mean even the JDK without the code change also work in 8.0?
> 

Yes, for my case. I tried kernel 4.18.0-147.8.1.el8_1.x86_64 and it is also working.

**Comment by LosAngelesHK** [comment]
> We're having similar issue but with Netty SCTP client not getting notification that SCTP association was created.
> This is also RHEL 8.2 and JDK 8 from Oracle.
> For us downgrading kernel to version from RHEL 8.0 solved this issue.
> We are currently trying to pin point release where issue was introduced.

@pgrabowski Actually, I have tried to compile and execute my SCTP server application with JDK 8, JDK 11, JDK 13 and JDK 14 in RHEL 8.2. All trials have the same issue that SCTP channel is blocked at channelRegistered() state.

**Comment by pgrabowski** [comment]
> @pgrabowski For your information, I have tried to compile and execute my SCTP server application with JDK 8, JDK 11, JDK 13 and JDK 14 in RHEL 8.2. All trials have the same issue that SCTP channel is blocked at channelRegistered() state.

Try downgrading kernel to 4.18.0-147.8.1.el8_1.x86_64. yum -y install kernel-modules-extra-4.18.0-147.8.1.el8_1.x86_64 and reboot choosing correct kernel in grub

**Comment by pgrabowski** [comment]
@LosAngelesHK I did a test and confirm that your code works on old version of kernel.

`
[root@sdp-rhinoref01 TestSctpCliSvr]# uname -a
Linux sdp-rhinoref01 4.18.0-147.8.1.el8_1.x86_64 #1 SMP Wed Feb 26 03:08:15 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
[root@sdp-rhinoref01 TestSctpCliSvr]# cat /etc/redhat-release
Red Hat Enterprise Linux release 8.2 (Ootpa)
[root@sdp-rhinoref01 TestSctpCliSvr]# java TestSctpSvr 127.0.0.1 37000 &
[1] 12674
[root@sdp-rhinoref01 TestSctpCliSvr]# SCTP_INIT_MAXSTREAMS,in = 65535, out = 10
SCTP_INIT_MAXSTREAMS,in = 10, out = 20
java TestSctpCli 127.0.0.1 37001 127.0.0.1 37000
SO_RCVBUF = 65536
SO_SNDBUF = 65536
SCTP_DISABLE_FRAGMENTS = false
SCTP_INIT_MAXSTREAMS,in = 65535, out = 10
SCTP_INIT_MAXSTREAMS,in = 100, out = 200
after connect, SCTP_INIT_MAXSTREAMS from assoc,in = 20, out = 10
sending packet 0 of size 1000, retval = 1000
Client connection received from [/127.0.0.1:37001], association = sun.nio.ch.sctp.AssociationImpl@33909752[associationID:8, maxIn:10, maxOut:20]
after connect, SCTP_INIT_MAXSTREAMS from assoc,in = 10, out = 20
Received a messagesun.nio.ch.sctp.MessageInfoImpl@55f96302[Address: /127.0.0.1:37001, Association: sun.nio.ch.sctp.AssociationImpl@33909752[associationID:8, maxIn:10, maxOut:20], Assoc ID: 8, Bytes: 1000, Stream Number: 0, Complete: true, isUnordered: false]
sending packet 1 of size 1000, retval = 1000
sending packet 2 of size 1000, retval = 1000
sending packet 3 of size 1000, retval = 1000
sending packet 4 of size 1000, retval = 1000
sending packet 5 of size 1000, retval = 1000
sending packet 6 of size 1000, retval = 1000
^C[root@sdp-rhinoref01 TestSctpCliSvr]# java -version
java version "1.8.0_261"
Java(TM) SE Runtime Environment (build 1.8.0_261-b33)
Java HotSpot(TM) 64-Bit Server VM (build 25.261-b33, mixed mode)
`

Result from new kernel.

`
[root@sdp-t-rhino01 TestSctpCliSvr]# uname -a
Linux sdp-t-rhino01 4.18.0-193.6.3.el8_2.x86_64 #1 SMP Mon Jun 1 20:24:55 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
[root@sdp-t-rhino01 TestSctpCliSvr]# java -version
java version "1.8.0_261"
Java(TM) SE Runtime Environment (build 1.8.0_261-b33)
Java HotSpot(TM) 64-Bit Server VM (build 25.261-b33, mixed mode)
[root@sdp-t-rhino01 TestSctpCliSvr]# cat /etc/redhat-
cat: /etc/redhat-: No such file or directory
[root@sdp-t-rhino01 TestSctpCliSvr]# cat /etc/redhat-release
Red Hat Enterprise Linux release 8.2 (Ootpa)
[root@sdp-t-rhino01 TestSctpCliSvr]# java TestSctpSvr 127.0.0.1 37000 &
[1] 1205209
[root@sdp-t-rhino01 TestSctpCliSvr]# SCTP_INIT_MAXSTREAMS,in = 65535, out = 10
SCTP_INIT_MAXSTREAMS,in = 10, out = 20

[root@sdp-t-rhino01 TestSctpCliSvr]# java TestSctpCli 127.0.0.1 37001 127.0.0.1 37000
SO_RCVBUF = 65536
SO_SNDBUF = 65536
SCTP_DISABLE_FRAGMENTS = false
SCTP_INIT_MAXSTREAMS,in = 65535, out = 10
SCTP_INIT_MAXSTREAMS,in = 100, out = 200
Exception in thread "main" java.lang.NullPointerException
        at TestSctpCli.main(TestSctpCli.java:42)
Client connection received from [/127.0.0.1:37001], association = null
Exception in thread "main" java.lang.NullPointerException
        at TestSctpSvr.main(TestSctpSvr.java:41)
[1]+  Exit 1                  java TestSctpSvr 127.0.0.1 37000
`

@normanmaurer I believe this ticket is not related to Netty but Red Hat Linux kernel.

**Comment by pgrabowski** [comment]
FYI there will be a fix in Red Hat Linux kernel for this issue.

**Comment by poekmorgan** [comment]
Doesn't look like there is a change on kernel. Or, I am confused? sorry, I am not familiar on it

**Comment by LosAngelesHK** [comment]
@pgrabowski Thanks for your information. In your issue report, it stated that "Issue is gone when one of IP addresses is removed from dispatcher.list". Could you provide more details about this solution?

**Comment by pgrabowski** [comment]
> Doesn't look like there is a change on kernel. Or, I am confused? sorry, I am not familiar on it

New kernel package is not yet officially released by Red Hat, test version is working as expected. You just need to wait.

> @pgrabowski Thanks for your information. In your issue report, it stated that "Issue is gone when one of IP addresses is removed from dispatcher.list". Could you provide more details about this solution?

I assume you are referring to Kamailio issue:) I think there is a difference in how Java and Kamailio handle SCTP association events received from Linux kernel so this solution is not applicable for Java. This is fix from upstream https://lkml.org/lkml/2020/6/9/972

**Comment by poekmorgan** [comment]
it sounds like a good news.

**Comment by poekmorgan** [comment]
@pgrabowski are you saying that you have tested with the new kernel?

**Comment by pgrabowski** [comment]
> @pgrabowski are you saying that you have tested with the new kernel?

Yes, I have tested a kernel with mentioned commit added on top. Official release should happen before RHEL 8.3.

**Comment by normanmaurer** [comment]
Let me close this one as it seems like it is a kernel bug and will be addressed

**Comment by pgrabowski** [comment]
Sorry for late reply. I wasn't able to retest it yet with bug fix kernel but this is my assumption that this was issue introduced in kernel for RHEL 8.2.
